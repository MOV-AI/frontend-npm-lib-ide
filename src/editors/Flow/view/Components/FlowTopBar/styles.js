import { makeStyles } from "@material-ui/styles";

export const buttonStyles = makeStyles((_theme) => ({
  buttonPill: {
    borderRadius: "99px",
    padding: "4px",
  },
}));

export const flowTopBarStyles = makeStyles((theme) => ({
  flowLink: {
    textDecoration: "underline",
    cursor: "pointer",
  },
  defaultRobot: {
    fontWeight: "bold",
  },
  formControl: {
    margin: theme.spacing(1),
    minWidth: 170,
    maxWidth: 350,
    "& i": {
      marginRight: 15,
    },
  },
  whichFlowText: {
    marginLeft: theme.spacing(5),
    fontSize: "15px",
    flexGrow: 1,
  },
  searchFlowArea: {
    marginRight: theme.spacing(3),
  },
  searchInputText: {
    width: "400px",
  },
  searchPopup: {
    width: "fit-content",
  },
  visualizationToggle: {
    marginRight: "10px",
  },
  grow: {
    flexGrow: 1,
    display: "flex",
    alignItems: "center",
  },
  treeIcon: {
    fontSize: "1.2rem",
  },
}));
