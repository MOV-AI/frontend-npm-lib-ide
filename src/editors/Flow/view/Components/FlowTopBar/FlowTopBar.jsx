import React, {
  forwardRef,
  useCallback,
  useEffect,
  useState,
  useMemo,
  useRef,
} from "react";
import PropTypes from "prop-types";
import { i18n } from "@mov-ai/mov-fe-lib-react";
import { RobotManager, CONSTANTS } from "@mov-ai/mov-fe-lib-core";
import {
  Typography,
  Tooltip,
  Button,
  CircularProgress,
} from "@material-ui/core";
import GrainIcon from "@material-ui/icons/Grain";
import FormControl from "@material-ui/core/FormControl";
import Select from "@material-ui/core/Select";
import MenuItem from "@material-ui/core/MenuItem";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import PlayArrowIcon from "@material-ui/icons/PlayArrow";
import StopIcon from "@material-ui/icons/Stop";
import { ToggleButton, ToggleButtonGroup } from "@material-ui/lab";
import Workspace from "../../../../../utils/Workspace";
import {
  SCOPES,
  PLUGINS,
  ALERT_SEVERITIES,
  ROBOT_BLACKLIST,
} from "../../../../../utils/Constants";
import {
  ERROR_MESSAGES,
  SUCCESS_MESSAGES,
} from "../../../../../utils/Messages";
import { defaultFunction } from "../../../../../utils/Utils";
import { FLOW_VIEW_MODE } from "../../Constants/constants";
import useNodeStatusUpdate from "./hooks/useNodeStatusUpdate";

import { buttonStyles, flowTopBarStyles } from "./styles";
import FlowSearch from "./FlowSearch";

const BACKEND_CALLBACK_NAME = "backend.FlowTopBar";

const ButtonTopBar = forwardRef((props, ref) => {
  const { disabled, onClick, children, testId = "input_top-bar" } = props;
  const classes = buttonStyles();
  return (
    <Button
      data-testid={testId}
      ref={ref}
      size="small"
      color="primary"
      variant="contained"
      disabled={disabled}
      className={classes.buttonPill}
      onClick={onClick}
    >
      {children}
    </Button>
  );
});

ButtonTopBar.propTypes = {
  disabled: PropTypes.bool,
  testId: PropTypes.string,
  onClick: PropTypes.func,
  children: PropTypes.node,
};

const FlowTopBar = (props) => {
  // Props
  const {
    call,
    alert,
    scope,
    loading,
    mainInterface,
    id,
    name,
    onRobotChange,
    onViewModeChange,
    viewMode,
    searchProps,
    confirmationAlert,
    canRun,
    robotSelected: robotSelect,
  } = props;
  // State hooks
  const [actionLoading, setActionLoading] = useState(false);
  const [robotSelected, setRobotSelected] = useState(robotSelect || "");

  // Sync robotSelected state with robotSelect prop changes
  useEffect(() => {
    console.log("robotSelect prop changed:", robotSelect);
    setRobotSelected(robotSelect || "");
  }, [robotSelect]);

  const [robotList, setRobotList] = useState({});

  // Other hooks
  const classes = flowTopBarStyles();
  const { robotSubscribe, robotUnsubscribe, getFlowPath, robotStatus } =
    useNodeStatusUpdate(props, robotSelected, viewMode);

  // Refs
  const buttonDOMRef = useRef();
  const helperRef = useRef();
  const isMounted = useRef();
  const flowInstanceRef = useRef();
  const requestedSuccessfulActionRef = useRef("");

  const [isActive, setIsActive] = useState(
    getFlowPath() === robotStatus.activeFlow,
  );

  // Managers Memos
  const robotManager = useMemo(() => new RobotManager(), []);
  const workspaceManager = useMemo(() => new Workspace(), []);

  //========================================================================================
  /*                                                                                      *
   *                                    Private Methods                                   *
   *                                                                                      */
  //========================================================================================

  /**
   * @private Change robot subscriber (on change of selected robot)
   * @param {String} robotId : New selected robot
   */
  const changeRobotSubscriber = useCallback(
    (robotId) => {
      robotUnsubscribe();
      robotSubscribe(robotId);
    },
    [robotSubscribe, robotUnsubscribe],
  );

  /**
   * @private Get store helper to use cloud functions
   */
  const initStoreHelper = useCallback(() => {
    return call(
      PLUGINS.DOC_MANAGER.NAME,
      PLUGINS.DOC_MANAGER.CALL.GET_STORE,
      scope,
    ).then((store) => {
      helperRef.current = store.helper;
      return store.helper;
    });
  }, [call, scope]);

  /**
   * @private Get store helper to use cloud functions
   */
  const initFlowInstance = useCallback(() => {
    return call(PLUGINS.DOC_MANAGER.NAME, PLUGINS.DOC_MANAGER.CALL.READ, {
      scope,
      name,
    }).then((document) => {
      flowInstanceRef.current = document;
      return document;
    });
  }, [call, scope, name]);

  /**
   * @private Init selected robot
   * @param {string} robotId : Robot ID
   */
  const initSelectedRobot = useCallback(
    (robotId) => {
      setRobotSelected(robotId);
      robotSubscribe(robotId);
      onRobotChange(robotId);
    },
    [onRobotChange, robotSubscribe],
  );

  /**
   * Get running default robot
   * @param {String} currentRobot : current selected robot id
   */
  const getRunningRobot = useCallback(
    (currentRobot, robots) => {
      // If there's a currently selected robot in local storage
      if (currentRobot) initSelectedRobot(currentRobot);
      // Call callback to get default robot
      const helper = helperRef.current;
      helper
        .getDefaultRobot()
        .then((robotId) => {
          // Update default robot in state and set as selected if there's none
          if (robotId) {
            setRobotList((prevState) => {
              return {
                ...prevState,
                [robotId]: { ...prevState[robotId], isDefault: true },
              };
            });
            // Update Flow selected robot if none is selected yet
            let robotToSelect = currentRobot;
            if (!currentRobot || !(currentRobot in robots)) {
              workspaceManager.setSelectedRobot(robotId);
              initSelectedRobot(robotId);
              robotToSelect = robotId;
            }
            // Set selected robot
            setRobotSelected(robotToSelect);
          }
        })
        .catch((err) => {
          console.warn("getRunningRobot error", err);
          alert({
            message: i18n.t(ERROR_MESSAGES.ERROR_RUNNING_SPECIFIC_CALLBACK, {
              callbackName: BACKEND_CALLBACK_NAME,
            }),
            severity: ALERT_SEVERITIES.ERROR,
          });
        });
    },
    [initSelectedRobot, workspaceManager, alert],
  );

  //========================================================================================
  /*                                                                                      *
   *                                    React Lifecycle                                   *
   *                                                                                      */
  //========================================================================================

  /**
   * On Load Robot List
   * @param {object} robots : All robots load from robotManager
   */
  const onLoadRobotList = useCallback(
    (robots) => {
      console.log(
        "onLoadRobotList",
        robotSelected,
        workspaceManager.getSelectedRobot(),
      );
      const currentSelected =
        robotSelected || workspaceManager.getSelectedRobot();
      // Remove blacklisted robots
      Object.keys(robots).forEach((robotId) => {
        if (ROBOT_BLACKLIST.includes(robots[robotId].RobotName))
          delete robots[robotId];
      });
      // Update state hooks
      setRobotList(robots);
      // Get running Robot
      getRunningRobot(currentSelected, robots);
    },
    [getRunningRobot, workspaceManager],
  );

  /**
   * On component mount
   */
  useEffect(() => {
    // Load store loader and robot list
    initStoreHelper().then(() => {
      robotManager.getAll(onLoadRobotList);
    });
    // Load flow instance
    initFlowInstance();
    // Set isMounted
    isMounted.current = true;

    /**
     * On component unmount
     */
    return () => {
      robotUnsubscribe();
      // Unmount
      isMounted.current = false;
    };
  }, [
    initFlowInstance,
    initStoreHelper,
    onLoadRobotList,
    robotManager,
    robotUnsubscribe,
  ]);

  /**
   * Finish actionLoading when there's an update on activeFlow
   */
  useEffect(() => {
    setActionLoading(false);
    setIsActive(getFlowPath() === robotStatus.activeFlow);

    if (requestedSuccessfulActionRef.current) {
      alert({
        message: i18n.t(SUCCESS_MESSAGES.SUCCESSFUL_FLOW_ACTION, {
          action: requestedSuccessfulActionRef.current,
        }),
        severity: ALERT_SEVERITIES.SUCCESS,
      });
    }

    setActionLoading(false);
    requestedSuccessfulActionRef.current = "";
  }, [robotStatus.activeFlow, getFlowPath, setActionLoading, alert]);

  //========================================================================================
  /*                                                                                      *
   *                                        Helper                                        *
   *                                                                                      */
  //========================================================================================

  /**
   * Check if can start flow
   * @param {string} action : One of values: "START" or "STOP"
   * @returns {boolean} True if can run flow and False otherwise
   */
  const canRunFlow = useCallback(
    (action) => {
      const graph = mainInterface.current?.graph;
      // let's validate flow before continuing
      graph?.validateFlow();

      const warnings = graph.warnings || [];
      const warningsVisibility = graph.warningsVisibility;
      const runtimeWarnings = warnings.filter((wn) => wn.isRuntime);
      runtimeWarnings.forEach((warning) => {
        graph.setPermanentWarnings(warning);
        if (!warningsVisibility)
          alert({
            message: runtimeWarnings[0].message,
            severity: runtimeWarnings[0].type,
          });
      });
      return !(runtimeWarnings.length && action === "START");
    },
    [alert, mainInterface],
  );

  //========================================================================================
  /*                                                                                      *
   *                                        Events                                        *
   *                                                                                      */
  //========================================================================================

  /**
   * Handle change of selected robot
   * @param {Event} event : Change native event
   */
  const handleChangeRobot = useCallback(
    (event) => {
      setRobotSelected((prevState) => {
        const robotId = event.target.value;
        if (robotId !== prevState) {
          // Set in local storage
          workspaceManager.setSelectedRobot(robotId);
          // Change subscriber and update state
          changeRobotSubscriber(robotId);
          onRobotChange(robotId);
          return robotId;
        }
        return prevState;
      });
    },
    [changeRobotSubscriber, onRobotChange, workspaceManager],
  );

  /**
   * Send action to start robot
   * @param {*} action
   * @returns To avoid starting flow if flow is not eligible to start
   */
  const sendActionToRobot = useCallback(
    async (action, flowPath) => {
      const canStart = canRunFlow(action);
      if (!canStart) return;

      setActionLoading(true);
      try {
        const res = helperRef.current.sendToRobot({
          action,
          flowPath: flowPath || getFlowPath(),
          robotId: robotSelected,
        });

        if (!res || !isMounted.current) {
          alert({
            message: i18n.t("FailedFlowAction", {
              action: i18n.t(action.toLowerCase()),
            }),
            severity: ALERT_SEVERITIES.ERROR,
          });
          return;
        }

        requestedSuccessfulActionRef.current = action;
      } catch (err) {
        console.warn("Error sending action to robot", err);
        alert({
          message: i18n.t(ERROR_MESSAGES.ERROR_RUNNING_SPECIFIC_CALLBACK, {
            callbackName: BACKEND_CALLBACK_NAME,
          }),
          severity: ALERT_SEVERITIES.ERROR,
        });
      }
    },
    [alert, canRunFlow, getFlowPath, robotSelected],
  );

  /**
   * Handle Start flow button click
   */
  const handleStartFlow = useCallback(
    (saveResponse) => {
      const flowUrl = saveResponse?.model?.getUrl();

      // Start Flow if there's no active flow
      if (robotStatus.activeFlow === "") {
        sendActionToRobot("START", flowUrl);
      } else {
        // Confirmation alert if Another flow is running!
        const title = i18n.t("AnotherFlowRunningConfirmationTitle");
        const message = i18n.t("AnotherFlowRunningConfirmationMessage", {
          robotName: robotList[robotSelected].RobotName,
          activeFlow: robotStatus.activeFlow,
          id: id,
        });
        confirmationAlert({
          title,
          message,
          submitText: i18n.t("Run"),
          onSubmit: () => sendActionToRobot("START", flowUrl),
        });
      }
    },
    [
      robotStatus.activeFlow,
      sendActionToRobot,
      robotList,
      robotSelected,
      id,
      confirmationAlert,
    ],
  );

  /**
   * Handle Stop flow button click
   */
  const handleStopFlow = useCallback(() => {
    sendActionToRobot("STOP");
  }, [sendActionToRobot]);

  /**
   * Handle Save before starting a flow
   */
  const handleSaveBeforeStart = useCallback(() => {
    const isDirty = flowInstanceRef.current?.getDirty();
    if (isDirty) {
      call(
        PLUGINS.DOC_MANAGER.NAME,
        PLUGINS.DOC_MANAGER.CALL.SAVE,
        {
          scope,
          name,
        },
        handleStartFlow,
      );
    } else {
      handleStartFlow();
    }
  }, [call, handleStartFlow, name, scope]);

  /**
   * Handle view mode change : default view to tree view
   * @param {Event} event : Event change
   * @param {string} newViewMode : New value
   * @returns
   */
  const handleViewModeChange = useCallback(
    (_event, newViewMode) => {
      onViewModeChange(newViewMode);
    },
    [onViewModeChange],
  );

  //========================================================================================
  /*                                                                                      *
   *                                        Render                                        *
   *                                                                                      */
  //========================================================================================

  /**
   * Render Start button content
   * @returns {ReactElement} CircularProgress to indicate actionLoading or play icon with tooltip
   */
  const renderStartButton = useCallback(() => {
    // Render circular progress if actionLoading
    return actionLoading ? (
      <CircularProgress size={25} color="inherit" />
    ) : (
      <Tooltip title={i18n.t("StartFlow")}>
        <>
          <PlayArrowIcon /> {i18n.t("SaveAndRun")}
        </>
      </Tooltip>
    );
  }, [actionLoading]);

  /**
   * Render Stop button content
   * @returns {ReactElement} CircularProgress to indicate actionLoading or Stop icon with tooltip
   */
  const renderStopButton = useCallback(() => {
    // Render circular progress if actionLoading
    return actionLoading ? (
      <CircularProgress size={25} color="inherit" />
    ) : (
      <Tooltip title={i18n.t("StopFlow")}>
        <StopIcon />
      </Tooltip>
    );
  }, [actionLoading]);

  return (
    <AppBar
      data-testid="section-flow-top-bar"
      position="static"
      color="inherit"
    >
      <Toolbar variant="dense">
        <Typography component="div" className={classes.grow}>
          <FormControl className={classes.formControl}>
            <Select
              inputProps={{ "data-testid": "input_change-robot" }}
              id="robot-selector"
              value={robotSelected}
              startAdornment={<i className="icon-Happy"></i>}
              onChange={handleChangeRobot}
            >
              {Object.keys(robotList).map((robotId) => {
                const isDefaultRobot = robotList[robotId].isDefault;
                return (
                  <MenuItem
                    key={`robotList-${robotId}`}
                    value={robotId}
                    className={isDefaultRobot ? classes.defaultRobot : ""}
                  >
                    {robotList[robotId].RobotName}
                  </MenuItem>
                );
              })}
            </Select>
          </FormControl>
          {isActive ? (
            <ButtonTopBar
              testId="input_stop-flow"
              ref={buttonDOMRef}
              disabled={actionLoading}
              onClick={handleStopFlow}
            >
              {renderStopButton()}
            </ButtonTopBar>
          ) : (
            <ButtonTopBar
              testId="input_save-before-start"
              ref={buttonDOMRef}
              disabled={!robotStatus.isOnline || actionLoading || !canRun}
              onClick={handleSaveBeforeStart}
            >
              {renderStartButton()}
            </ButtonTopBar>
          )}
        </Typography>
        <Typography
          data-testid="section_view-search"
          component="div"
          className={classes.searchFlowArea}
        >
          <FlowSearch {...searchProps} />
        </Typography>
        <Typography
          data-testid="section_view-mode-toggle"
          component="div"
          className={classes.visualizationToggle}
        >
          <ToggleButtonGroup
            exclusive
            size="small"
            value={viewMode}
            onChange={handleViewModeChange}
          >
            <ToggleButton
              data-testid="input_default-flow"
              value={FLOW_VIEW_MODE.default}
              disabled={loading}
            >
              <Tooltip title={i18n.t("DefaultFlowView")}>
                <GrainIcon fontSize="small" />
              </Tooltip>
            </ToggleButton>
            <ToggleButton
              data-testid="input_tree-view-flow"
              value={FLOW_VIEW_MODE.treeView}
              disabled={loading}
            >
              <Tooltip title={i18n.t("TreeView")}>
                <i className={`icon-tree ${classes.treeIcon}`}></i>
              </Tooltip>
            </ToggleButton>
          </ToggleButtonGroup>
        </Typography>
      </Toolbar>
    </AppBar>
  );
};

FlowTopBar.propTypes = {
  id: PropTypes.string,
  scope: PropTypes.string,
  viewMode: PropTypes.string,
  name: PropTypes.string,
  loading: PropTypes.bool,
  canRun: PropTypes.bool,
  mainInterface: PropTypes.object,
  call: PropTypes.func,
  confirmationAlert: PropTypes.func,
  alert: PropTypes.func,
  nodeStatusUpdated: PropTypes.func,
  onViewModeChange: PropTypes.func,
  onStartStopFlow: PropTypes.func,
  onRobotChange: PropTypes.func,
  openFlow: PropTypes.func,
  workspace: PropTypes.string,
  type: PropTypes.string,
  version: PropTypes.string,
  searchProps: PropTypes.shape({
    visible: PropTypes.bool,
    options: PropTypes.arrayOf(PropTypes.object),
    onChange: PropTypes.func,
    onEnabled: PropTypes.func,
    onDisabled: PropTypes.func,
  }),
};

FlowTopBar.defaultProps = {
  openFlow: () => defaultFunction("openFlow"),
  onRobotChange: () => defaultFunction("onRobotChange"),
  onViewModeChange: () => defaultFunction("onViewModeChange"),
  onStartStopFlow: () => defaultFunction("onStartStopFlow"),
  nodeStatusUpdated: () => defaultFunction("nodeStatusUpdated"),
  workspace: CONSTANTS.GLOBAL_WORKSPACE,
  type: SCOPES.FLOW,
  version: "__UNVERSIONED__",
};

export default FlowTopBar;
