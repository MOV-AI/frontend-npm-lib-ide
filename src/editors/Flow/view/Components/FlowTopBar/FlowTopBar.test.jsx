/* eslint-env jest */

jest.mock("../../../../../utils/Workspace", () => {
  const workspaceMock = {
    getSelectedRobot: jest.fn().mockReturnValue(undefined),
    setSelectedRobot: jest.fn().mockImplementation((id) => {
      workspaceMock.getSelectedRobot.mockReturnValue(id);
    }),
  };
  const WorkspaceFactory = jest.fn().mockImplementation(() => workspaceMock);
  WorkspaceFactory.__workspaceMock = workspaceMock;
  return { __esModule: true, default: WorkspaceFactory };
});

import React from "react";
import { render, screen, within, waitFor, act } from "@testing-library/react";
import { i18n } from "@mov-ai/mov-fe-lib-react";
import { Rest } from "@mov-ai/mov-fe-lib-core";
import userEvent from "@testing-library/user-event";
import WorkspaceFactory from "../../../../../utils/Workspace";

import FlowTopBar from "./FlowTopBar";

jest.mock("@mov-ai/mov-fe-lib-core", () => ({
  RobotManager: jest.fn().mockImplementation(() => ({
    getAll: jest.fn((cb) =>
      cb({ r1: { RobotName: "Robot One" }, r2: { RobotName: "Robot Two" } }),
    ),
  })),
  CONSTANTS: { GLOBAL_WORKSPACE: "global" },
  Rest: {
    post: jest.fn().mockResolvedValue({ success: true }),
  },
}));

jest.mock("./hooks/useNodeStatusUpdate", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("./styles", () => ({
  flowTopBarStyles: () => ({}),
  buttonStyles: () => ({}),
}));

describe("FlowTopBar", () => {
  let callMock,
    helper,
    onRobotChange,
    onViewModeChange,
    confirmationAlert,
    mainInterface;

  beforeEach(() => {
    jest.clearAllMocks();

    WorkspaceFactory.__workspaceMock.getSelectedRobot.mockReturnValue(
      undefined,
    );

    helper = {
      getDefaultRobot: jest.fn().mockResolvedValue("r1"),
      sendToRobot: jest.fn().mockResolvedValue(true),
    };

    callMock = jest.fn((plugin, method, args, cb) => {
      if (method === "getStore") return Promise.resolve({ helper });
      if (method === "read")
        return Promise.resolve({
          getDirty: jest.fn().mockReturnValue(false),
          model: { getUrl: jest.fn().mockReturnValue("path1") },
        });
      if (method === "save") {
        cb({ model: { getUrl: jest.fn().mockReturnValue("path1") } });
        return Promise.resolve();
      }
      return Promise.resolve();
    });

    require("./hooks/useNodeStatusUpdate").default.mockReturnValue({
      robotSubscribe: jest.fn(),
      robotUnsubscribe: jest.fn(),
      getFlowPath: jest.fn(() => "path1"),
      robotStatus: { activeFlow: "", isOnline: true },
    });

    onRobotChange = jest.fn();
    onViewModeChange = jest.fn();
    confirmationAlert = jest.fn();
    mainInterface = {
      current: {
        graph: {
          validateFlow: jest.fn(),
          warnings: [],
          warningsVisibility: true,
          setPermanentWarnings: jest.fn(),
        },
      },
    };
  });

  async function setup(props = {}) {
    let container;
    await act(async () => {
      const result = render(
        <FlowTopBar
          call={callMock}
          alert={jest.fn()}
          scope="flow"
          loading={false}
          mainInterface={mainInterface}
          id="flow1"
          name="flowName"
          onRobotChange={onRobotChange}
          onViewModeChange={onViewModeChange}
          viewMode="default"
          searchProps={{
            visible: true,
            options: [],
            onChange: jest.fn(),
            onEnabled: jest.fn(),
            onDisabled: jest.fn(),
          }}
          confirmationAlert={confirmationAlert}
          canRun={true}
          {...props}
        />,
      );
      container = result.container;
    });

    await waitFor(() =>
      expect(callMock).toHaveBeenCalledWith("docManager", "getStore", "flow"),
    );

    await waitFor(() => expect(helper.getDefaultRobot).toHaveBeenCalled());

    return container;
  }

  it("renders the two view-mode buttons and the robot selector", async () => {
    await setup();
    expect(screen.getByTestId("input_tree-view-flow")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /Robot/i })).toBeInTheDocument();
  });

  it("changes view mode", async () => {
    await setup();

    act(() => {
      userEvent.click(screen.getByTestId("input_tree-view-flow"));
    });

    expect(onViewModeChange).toHaveBeenCalledWith("treeView");
  });

  it("selects a new robot", async () => {
    await setup();

    act(() => {
      userEvent.click(screen.getByRole("button", { name: "Robot One" }));
    });

    const listbox = await screen.findByRole("listbox", { hidden: true });

    expect(listbox).toBeInTheDocument();

    const option = within(listbox).getByRole("option", {
      name: "Robot Two",
      hidden: true,
    });

    await act(async () => userEvent.click(option));

    const workspaceMock = WorkspaceFactory.__workspaceMock;

    expect(workspaceMock.setSelectedRobot).toHaveBeenCalledWith("r2");
    expect(workspaceMock.getSelectedRobot()).toBe("r2");
    expect(onRobotChange).toHaveBeenCalledWith("r2");
  });

  it("starts flow", async () => {
    await setup();

    act(() => {
      userEvent.click(screen.getByTestId("input_save-before-start"));
    });

    await waitFor(() =>
      expect(helper.sendToRobot).toHaveBeenCalledWith({
        action: "START",
        flowPath: "path1",
        robotId: "r1",
      }),
    );
  });

  it("stops flow", async () => {
    require("./hooks/useNodeStatusUpdate").default.mockReturnValue({
      robotSubscribe: jest.fn(),
      robotUnsubscribe: jest.fn(),
      getFlowPath: jest.fn(() => "path1"),
      robotStatus: { activeFlow: "path1", isOnline: true },
    });

    await setup();

    act(() => {
      userEvent.click(screen.getByTestId("input_stop-flow"));
    });

    await waitFor(() =>
      expect(helper.sendToRobot).toHaveBeenCalledWith({
        action: "STOP",
        flowPath: "path1",
        robotId: "r1",
      }),
    );
  });

  it("prompts when another flow is running", async () => {
    require("./hooks/useNodeStatusUpdate").default.mockReturnValue({
      robotSubscribe: jest.fn(),
      robotUnsubscribe: jest.fn(),
      getFlowPath: jest.fn(() => "path1"),
      robotStatus: { activeFlow: "otherFlow", isOnline: true },
    });

    await setup();

    await act(async () =>
      userEvent.click(screen.getByTestId("input_save-before-start")),
    );

    await waitFor(() =>
      expect(confirmationAlert).toHaveBeenCalledWith(
        expect.objectContaining({
          title: i18n.t("AnotherFlowRunningConfirmationTitle"),
          message: i18n.t("AnotherFlowRunningConfirmationMessage", {
            robotName: "Robot One",
            activeFlow: "otherFlow",
            id: "flow1",
          }),
          submitText: i18n.t("Run"),
          onSubmit: expect.any(Function),
        }),
      ),
    );
  });
});
