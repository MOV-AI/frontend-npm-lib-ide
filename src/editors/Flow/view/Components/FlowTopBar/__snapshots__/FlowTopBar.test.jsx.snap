// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowTopBar matches snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorInherit MuiPaper-elevation4"
    data-testid="section-flow-top-bar"
  >
    <div
      class="MuiToolbar-root MuiToolbar-dense MuiToolbar-gutters"
    >
      <div
        class="MuiTypography-root MuiTypography-body1"
      >
        <div
          class="MuiFormControl-root"
        >
          <div
            class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart"
          >
            <i
              class="icon-Happy"
            />
            <div
              aria-haspopup="listbox"
              aria-labelledby="robot-selector"
              class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedStart"
              id="robot-selector"
              role="button"
              tabindex="0"
            >
              Robot One
            </div>
            <input
              aria-hidden="true"
              class="MuiSelect-nativeInput"
              data-testid="input_change-robot"
              tabindex="-1"
              value="r1"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
          data-testid="input_save-before-start"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8 5v14l11-7z"
              />
            </svg>
             
            Save & Run
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="MuiTypography-root MuiTypography-body1"
        data-testid="section_view-search"
      >
        <div
          aria-expanded="false"
          class="MuiAutocomplete-root Mui-focused MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon"
          data-testid="input_search-flow-text"
          role="combobox"
        >
          <div
            class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiAutocomplete-inputRoot MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <input
                aria-autocomplete="list"
                aria-invalid="false"
                autocapitalize="none"
                autocomplete="off"
                class="MuiInputBase-input MuiInput-input MuiAutocomplete-input MuiAutocomplete-inputFocused MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd"
                data-scope="flowSearchInput"
                id="mui-36741"
                placeholder="Search node..."
                spellcheck="false"
                type="text"
                value=""
              />
              <div
                class="MuiAutocomplete-endAdornment"
              >
                <button
                  aria-label="Clear"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-clearIndicator"
                  tabindex="-1"
                  title="Clear"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
                <button
                  aria-label="Open"
                  class="MuiButtonBase-root MuiIconButton-root MuiAutocomplete-popupIndicator"
                  tabindex="-1"
                  title="Open"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M7 10l5 5 5-5z"
                      />
                    </svg>
                  </span>
                  <span
                    class="MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiTypography-root MuiTypography-body1"
        data-testid="section_view-mode-toggle"
      >
        <div
          class="MuiToggleButtonGroup-root"
          role="group"
        >
          <button
            aria-pressed="true"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal Mui-selected MuiToggleButton-sizeSmall"
            data-testid="input_default-flow"
            tabindex="0"
            type="button"
            value="default"
          >
            <span
              class="MuiToggleButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                focusable="false"
                title="Default Flow view"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 12c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zM6 8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12-8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm-4 8c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm4-4c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-4-4c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-4-4c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-pressed="false"
            class="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal MuiToggleButton-sizeSmall"
            data-testid="input_tree-view-flow"
            tabindex="0"
            type="button"
            value="treeView"
          >
            <span
              class="MuiToggleButton-label"
            >
              <i
                class="icon-tree undefined"
                title="Tree view"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </header>
</div>
`;
